### PR TITLE
Feat: typesense vector store

### PR DESCRIPTION
**Pull request by tomi-mercado**
Add an implementation of Typesense vector store with @cocomatias

Available methods:
- `fromDocuments`
- `fromTexts`
- `similaritySearch`
- `addDocuments`: Works like the others addDocuments, but update the data if `id` is repeated if use default `import`.
- `addDocumentsWithoutEmbedding`: Util to update documents without do their embeddings again. Will work for create add documents without embedding too.
- `deleteDocuments`: Receive an array of ids to delete
- `modifySearchParams`: Work for filter with `MultiSearchRequestSchema` in Typesense in next `similaritySearch`

All the metadata will be accesible and will be saved in the schema passing `columnNames.metadataColumnNames`

Has a default method to import data, but it's possible to override it with `import`.

### CODE REVIEW

File: langchain/src/vectorstores/typesense.ts [hunk 1]
```diff
@@ -0,0 +1,370 @@
...
+   */
+  async deleteDocuments(documentIds: string[]) {
+    await this.client
+      .collections(this.schemaName)
+      .documents()
+      .delete({
+        filter_by: `id:=${documentIds.join(",")}`,
+      });
+  }
+
```
Comment (dqbd) [inline langchain/src/vectorstores/typesense.ts:358]: I believe both `similaritySearchVectorWithScore` and `similaritySearchWithScore` can be used in other chains / retrievers. afaict the TypeSense API does not provide the similarity scores in their APIs, thus would it make sense to just use `1` as a score instead?

File: langchain/src/vectorstores/tests/typesense.test.ts [hunk 2]
```diff
@@ -0,0 +1,112 @@
...
+      vec: await embeddings.embedQuery("hello world 2"),
+    },
+  ];
+
+  expect(await vectorstore.documentsToTypesenseRecords(documents)).toEqual(
+    expected
+  );
+});
+
+test("typesenseRecordsToDocuments should return the correct langchain documents", async () => {
```
Comment (dqbd) [inline langchain/src/vectorstores/tests/typesense.test.ts:61]: Consider adding integration tests as well

**Review by dqbd — COMMENTED** [review]
Hi, thank you for your contribution! Some minor tweaks and questions..

Also, would it be possible to provide a documentation entry as well? Thanks!

File: langchain/package.json [hunk 3]
```diff
@@ -522,6 +525,7 @@
     "openai": "^3.2.0",
     "p-queue": "^6.6.2",
     "p-retry": "4",
+    "typesense": "^1.5.3",
```
[the remark this answers is not in the document]
Comment (tomi-mercado) [reply langchain/package.json]: I think is solved, can you check it?

File: langchain/package.json [hunk 4]
```diff
@@ -124,6 +124,9 @@
     "vectorstores/myscale.cjs",
     "vectorstores/myscale.js",
     "vectorstores/myscale.d.ts",
+    "vectorstores/typesense.cjs",
```
[the remark this answers is not in the document]
Comment (tomi-mercado) [reply langchain/package.json:127]: I think is solved, can you check it?

File: langchain/src/vectorstores/typesense.ts [hunk 5]
```diff
@@ -0,0 +1,370 @@
...
+    data: T[],
+    collectionName: string
+  ): Promise<void>;
+}
+
+/**
+ * Typesense vector store.
+ */
+export class Typesense extends VectorStore {
+  private client: Client;
```
[the remark this answers is not in the document]
Comment (tomi-mercado) [reply langchain/src/vectorstores/typesense.ts:65]: Fixed!

File: langchain/src/vectorstores/typesense.ts [hunk 6]
```diff
@@ -0,0 +1,370 @@
...
+      const chunk = data.slice(i, i + chunkSize);
+
+      try {
+        await typesenseClient
+          .collections<T>(collectionName)
+          .documents()
+          .import(chunk, { action: "emplace", dirty_values: "drop" });
+      } catch (error) {
+        console.log("Error importing data to typesense");
+        console.error(error);
```
[the remark this answers is not in the document]
Comment (tomi-mercado) [reply langchain/src/vectorstores/typesense.ts]: You're right, comes from another implementation so I had this mistake. Fixed! Good catch

File: langchain/src/vectorstores/typesense.ts [hunk 7]
```diff
@@ -0,0 +1,370 @@
...
+      numRetries: 3,
+      connectionTimeoutSeconds: TEN_MINUTES_IN_SECONDS,
+    });
+
+    const chunkSize = 2000;
+    for (let i = 0; i < data.length; i += chunkSize) {
+      const chunk = data.slice(i, i + chunkSize);
+
+      try {
+        await typesenseClient
```
[the remark this answers is not in the document]
Comment (tomi-mercado) [reply langchain/src/vectorstores/typesense.ts:137]: Can you check if is ok the implementation?

File: langchain/src/vectorstores/typesense.ts [hunk 8]
```diff
@@ -0,0 +1,370 @@
...
+   * @param data
+   * @param collectionName
+   */
+  async importToTypesense<T extends Record<string, any> = Record<string, any>>(
+    data: T[],
+    collectionName: string
+  ) {
+    const TEN_MINUTES_IN_SECONDS = 10 * 60;
+    const typesenseConfig = {
+      host: process.env.TYPESENSE_HOST,
```
[the remark this answers is not in the document]
Comment (tomi-mercado) [reply langchain/src/vectorstores/typesense.ts]: Thank you, fixed!

File: langchain/src/vectorstores/typesense.ts [hunk 9]
```diff
@@ -0,0 +1,370 @@
+import { Client } from "typesense";
+import { MultiSearchRequestSchema } from "typesense/lib/Typesense/MultiSearch.js";
```
[the remark this answers is not in the document]
Comment (tomi-mercado) [reply langchain/src/vectorstores/typesense.ts]: Fixed

File: langchain/src/vectorstores/typesense.ts (hunk 1 above)
Comment (tomi-mercado) [reply langchain/src/vectorstores/typesense.ts:369]: Has sense to me, I added the methods

File: langchain/src/vectorstores/typesense.ts [hunk 10]
```diff
@@ -0,0 +1,370 @@
...
+ */
+export class Typesense extends VectorStore {
+  private client: Client;
+  private schemaName: string;
+  private searchParams: MultiSearchRequestSchema;
+  private vectorColumnName: string;
+  private pageContentColumnName: string;
+  private metadataColumnNames: string[];
+  private import: (
+    data: Record<string, any>[],
```
[the remark this answers is not in the document]
Comment (tomi-mercado) [reply langchain/src/vectorstores/typesense.ts]: Fixed!

File: langchain/src/vectorstores/tests/typesense.test.ts (hunk 2 above)
Comment (tomi-mercado) [reply langchain/src/vectorstores/tests/typesense.test.ts:61]: I am not sure about how can I do it. I will ask for help in Discord, there was somebody that knows Langchain and maybe can guide me

**Comment by tomi-mercado** [comment]
@dqbd Thanks a lot for your review! Give me a few days to create a documentation and make better examples 👍

File: langchain/package.json (hunk 3 above)
[the remark this answers is not in the document]
Comment (tomi-mercado) [reply langchain/package.json]: Well, I see that the Vercel deploy failed so I think that is not well implemented, can you guide me about how can I fix it?

**Comment by johannschopplich** [comment]
I'm truly hyped for this PR. 🚀

**Comment by tomi-mercado** [comment]
I am really lost about how can I face the next steps. If someone could help me I will be glad to help to solve the things that I am missing

**Comment by tomi-mercado** [comment]
I will work tomorrow in the documentation. If someone could help me with integration tests, will be great!

**Comment by tomi-mercado** [comment]
Docs added!

**Comment by tomi-mercado** [comment]
> nfcampos

I removed `modifySearchParams` and use only `filter` argument of `similaritySearch` and `similaritySearchScore`

**Comment by tomi-mercado** [comment]
> This PR will need some updates to conform to the interface of the base class before being merged cc @dqbd Namely, addVectors support is missing (the usual way to implement this is to implement addVectors and then addDocuments is implemented by calling addVectors), and there is a non-standard `modifySearchParams` method which I don't understand the purpose of

Also I implemented addVectors too! Your guide was too helpful, thank you!

**Comment by nfcampos** [comment]
Thanks

**Pull request merged**